File: py/quick_sort.py
import csv
import time
import sys

sys.setrecursionlimit(10**8)

def read_dataset(filename):
    """Reads a slice of the dataset, returning a list of (int, str) tuples."""
    data = []
    with open(filename, 'r') as f:
        reader = csv.reader(f)
        for row in reader:
            if row:
                data.append((int(row[0]), row[1]))
    return data

def write_sorted_data(filename, sorted_data):
    """Writes the sorting steps to a file."""
    with open(filename, 'w', newline='') as f:
        writer = csv.writer(f)
        for item in sorted_data:
            writer.writerow(item)

def quick_sort(arr, low, high):
    """Performs quick sort on array."""
    if low < high:
        pi = partition(arr, low, high)
        quick_sort(arr, low, pi - 1)
        quick_sort(arr, pi + 1, high)

def median_of_three(arr, low, high):
    """
    Selects the median of the first, middle, and last elements as the pivot.
    Moves this pivot to the 'high' position for the partition function.
    """
    mid = (low + high) // 2
    
    # Sort low, mid, high to put median at mid
    if arr[low][0] > arr[mid][0]:
        arr[low], arr[mid] = arr[mid], arr[low]
    if arr[low][0] > arr[high][0]:
        arr[low], arr[high] = arr[high], arr[low]
    if arr[mid][0] > arr[high][0]:
        arr[mid], arr[high] = arr[high], arr[mid]
    
    # Place the pivot (median) at the high position
    arr[mid], arr[high] = arr[high], arr[mid]
    return arr[high][0] # Return the pivot value

def partition(arr, low, high):
    """
    Partitions the array using the chosen pivot.
    """
    pivot = median_of_three(arr, low, high) # Use median-of-three pivot
    i = (low - 1)
    
    for j in range(low, high):
        if arr[j][0] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i] # Swap elements

    arr[i + 1], arr[high] = arr[high], arr[i + 1] # Place pivot in correct position
    return i + 1

if __name__ == "__main__":
    dataset_filename = input("Enter dataset filename (e.g., dataset_1000000.csv): ")
    output_filename_base = dataset_filename.replace("dataset_", "quick_sort_").replace(".csv", "")

    data = read_dataset(dataset_filename)
    
    start_time = time.perf_counter() # Exclude I/O time 
    # Create a mutable copy for sorting as quick_sort modifies in-place
    data_to_sort = list(data)
    quick_sort(data_to_sort, 0, len(data_to_sort) - 1)
    end_time = time.perf_counter()

    running_time = end_time - start_time
    print(f"Quick sort running time: {running_time:.6f} seconds") # Print running time

    # Dynamically set the output filename based on input filename
    output_filename = f"{output_filename_base}.csv"
    write_sorted_data(output_filename, data_to_sort)
    print(f"Sorted data saved to {output_filename}")

File: CPP/quick_sort.cpp
#include <bits/stdc++.h> 
/*Libraries: iostream, fstream, string, vector, sstream
NOTE: PLEASE USE G++ COMPILER OR U MIGHT NOT RUN THIS PROGRAM*/
using namespace std;

string filename = "quick_sort";
int printout_count = 0;

struct Record {
    int key;
    string value;
};

vector<string> split(const string& str) {
    vector<string> parts;
    stringstream ss(str);
    string part;
    
    // Get the integer part before the first comma
    getline(ss, part, ',');
    parts.push_back(part); 
    
    // Get the string part (rest of the line)
    getline(ss, part);     
    parts.push_back(part);
    return parts;
}

vector<Record> read_dataset(const string& filename) {
    vector<Record> records;
    ifstream file(filename);
    string line;

    while (getline(file, line)) {
        if (line.empty()) continue;
        stringstream ss(line);
        string key_str, val;
        if (getline(ss, key_str, ',') && getline(ss, val)) {
            try {
                int key = stoi(key_str);
                records.push_back({key, val});
            } catch (...) {
                cerr << "Warning: Skipping malformed line: " << line << endl;
            }
        }
    }
    return records;
}

void save_dataset(const vector<Record>& dataset, const string& output_filename) {
    ofstream file(output_filename);
    if (!file.is_open()) {
        cerr << "Error: Could not open output file " << output_filename << endl;
        return;
    }

    for (size_t i = 0; i < dataset.size(); ++i) {
        file << dataset[i].key << "," << dataset[i].value;
        if (i < dataset.size() - 1) file << '\n';
    }
}

int partition(vector<Record>& dataset, int low, int high) {
    // Pivot is the last element's integer value
    int pivot_val = dataset[high].key;
    int i = (low - 1); // Index of smaller element

    for (int j = low; j < high; ++j) {
        // If current element is smaller than or equal to pivot
        if (dataset[j].key <= pivot_val) {
            i++; // Increment index of smaller element
            // Swap elements
            swap(dataset[i], dataset[j]);
        }
    }
    // Swap pivot to its correct position
    swap(dataset[i + 1], dataset[high]);
    return i + 1; // Return the partitioning index
}

void quick_sort(vector<Record>& dataset, int low, int high) {
    while (low < high) {
        int pi = partition(dataset, low, high);
        
        // Recurse on smaller half first to reduce stack depth
        if (pi - low < high - pi) {
            quick_sort(dataset, low, pi - 1);
            low = pi + 1; // Tail call optimization
        } else {
            quick_sort(dataset, pi + 1, high);
            high = pi - 1; // Tail call optimization
        }
    }
}


int main() {
    string dataset_input_filename;
    cout << "Enter dataset filename (e.g., dataset_1000000.csv): ";
    cin >> dataset_input_filename;

    // Read the dataset
    vector<Record> dataset = read_dataset(dataset_input_filename);
	
    if (dataset.empty()) {
        cerr << "No data read or file not found. Exiting." << endl;
        return 1;
    }

    // Prepare output filename (e.g., quick_sort_1000000.csv)
    // Find the position of "dataset_" and ".csv" to dynamically create the output filename
    string output_filename = dataset_input_filename;
    size_t dataset_prefix_pos = output_filename.find("dataset_");
    if (dataset_prefix_pos != string::npos) {
        output_filename.replace(dataset_prefix_pos, string("dataset_").length(), "quick_sort_");
    } else {
        // Fallback if "dataset_" prefix is not found
        output_filename = "quick_sort_" + output_filename;
    }

    cout << "Dataset size: " << dataset.size() << " elements." << endl;

    // Measure running time
    auto start_timer = chrono::high_resolution_clock::now();

    quick_sort(dataset, 0, dataset.size() - 1); // Call quick_sort on the entire dataset

    auto end_timer = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end_timer - start_timer);
    cout << "Quick sort running time: " << duration.count() << " ms" << endl;

    // Save the sorted dataset
    save_dataset(dataset, output_filename);
    cout << "Sorted data saved to " << output_filename << endl;

    return 0;
}